src/iop/useless.c

```diff
@@ -232,60 +290,185 @@ void cleanup_global(dt_iop_module_so_t *module)
+  // be done in gui_update (or gui_changed or individual widget callbacks)
+  // if any default values for (slider) widgets or options (in comboboxes) depend on the
+  // type of picture, then the widgets have to be updated in reload_params.
+  dt_iop_useless_gui_data_t *g = IOP_GUI_ALLOC(useless);
+
+  // if the first widget is created useing a _from_params call, self->widget does not have to
+  // be explicitly initialised as a new virtical box will be created automatically.
+  self->widget = gtk_box_new(GTK_ORIENTATION_VERTICAL, DT_BAUHAUS_SPACE);
+
+  // linking a slider to an int will make it take only whole numbers (step=1)
+  // the new slider is added to self->widget
+  g->scale = dt_bauhaus_slider_from_params(self, "checker_scale");
+
+  // if the field name should be used as label, it does not need a $DESCRIPTION
+  // mark it for translation here using N_()
+  // a colorpicker can be attached to a slider, as here, or put standalone in a box.
+  // when a color is picked, color_picker_apply is called with either the slider or the
+  // button that triggered it.
+  g->factor = dt_color_picker_new(self, DT_COLOR_PICKER_AREA,
+              dt_bauhaus_slider_from_params(self, N_("factor"));
+  // the initial slider range can be reduced from the introspection $MIN - $MAX
+  dt_bauhaus_slider_set_soft_range(g->scale, 0.5f, 1.5f);
+  // the default step is range/100, but can be changed here
+  dt_bauhaus_slider_set_step(g->grey, .1);
+  // additional parameters determine how the value will be shown.
+  dt_bauhaus_slider_set_format(g->grey, "%.2f %%");
+  dt_bauhaus_slider_set_digits(g->contrast, 2);
+  dt_bauhaus_slider_set_factor(g->contrast, -100.0f);
+  dt_bauhaus_slider_set_offset(g->contrast, 100.0f);
+  // tooltips explain the otherwise compact interface
+  gtk_widget_set_tooltip_text(g->grey, _("adjust factor"));
+
+  g->check = dt_bauhaus_toggle_from_params(self, "check");
+
+  // a combobox linked to struct field will be filled with the values and $DESCRIPTIONs
+  // in the struct definition, in the same order. the automatic callback will put the
+  // enum value, not the position in the combobox list, in the field.
+  g->method = dt_bauhaus_combobox_from_params(self, "method");
+
+  g->extra = dt_bauhaus_slider_new_with_range(self, -0.5, 0.5, .01, 0, 2);
+  dt_bauhaus_widget_set_label(g->scale, NULL, _("extra"));
+  gtk_box_pack_start(GTK_BOX(self->widget), GTK_WIDGET(g->extra), TRUE, TRUE, 0);
+  g_signal_connect(G_OBJECT(g->extra), "value-changed", G_CALLBACK(extra_callback), self);
 }
 
 void gui_cleanup(dt_iop_module_t *self)
 {
-  // nothing else necessary, gtk will clean up the slider.
-  free(self->gui_data);
-  self->gui_data = NULL;
+  // this only needs to be provided if gui_init allocates any memory or resources besides
+  // self->widget and gui_data_t. in that case IOP_GUI_FREE needs to be called to free the
+  // latter. the widget will be destroyed automatically.
```

Comment:
The term "the latter" in this sentence refers to "gui_data_t" but I'm not sure this was intentional.

Reply:
yes, that's what is meant. if you can think of a way to make that less ambiguous, I'll take it!

Reply:
Doesn't gui_cleanup also need to free the additional data/resource as well though (besides gui_data_t)? Isn't that the point of this function?

Reply:
Ah, but IOP_GUI_FREE can't do that...

Basically what I'm saying is, the default function (if an explicit one isn't provided here) takes care of gui_data_t (and widget gets destroyed anyway). If you _override_ the default, you have to do whatever you have to do, and _also_ call IOP_GUI_FREE to clean up gui_data_t.

Reply:
That sounds good. Say that.